datalad_next/iter_collections/annexworktree.py

```diff
@@ -0,0 +1,167 @@
+"""Report on the content of a Git-annex repository worktree
+
+The main functionality is provided by the :func:`iter_annexworktree()`
+function.
+"""
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from enum import Enum
+from pathlib import (
+    Path,
+    PurePath,
+)
+from typing import Generator
+
+from datalad_next.runners import StdOutCaptureGeneratorProtocol
+from datalad_next.runners.batch import (
+    annexjson_batchcommand,
+    annexline_batchcommand,
+)
+from datalad_next.runners.run import run
+
+from .gitworktree import (
+    GitWorktreeItem,
+    GitWorktreeFileSystemItem,
+    iter_gitworktree,
+    _lsfiles_line2props,
+    _mode_type_map,
+)
+
+lgr = logging.getLogger('datalad.ext.next.iter_collections.annexworktree')
+
+
+# The following protocol is part of the POC implementation. It allows a simple
+# `with run()`-commandline that returns `GitWorkTreeItem`-objects or
+# `GitWorkTreeFileSystemItem`-objects. In the real implementation we would
+# probably use `iter_gitworktree()`.
```

Comment:
This raises interesting points for me:

- yes, we would definitely use `iter_gitworktree()`, but `iter_gitworktree()` could be implemented this way for max alignment
- `GitLsFilesProtocol` should probably not sit on `StdOutCaptureGeneratorProtocol` directly, but on an intermediate protocol that implements the zero-byte-delimited-items output, or even delimited-items output. Right now the protocol assume that `ls-files` is called correctly but in the line
  ```
    run(['git', 'ls-files', '-z', '--stage', '--cached'], protocol_class=GitLsFilesProtocol...
  ```

  I cannot immediately see this requirement/implication.

- it seems we should/could aim for assembling a library of such common protocols

Reply:
> This raises interesting points for me:
> 
> * yes, we would definitely use `iter_gitworktree()`, but `iter_gitworktree()` could be implemented this way for max alignment

That's what I thought as well. I was tempted to do that. But I refrained, in order to focus on the annex-batch-related parts.

> * `GitLsFilesProtocol` should probably not sit on `StdOutCaptureProcessingGeneratorProtocol` directly, but on an intermediate protocol that implements the zero-byte-delimited-items output, or even delimited-items output. Right now the protocol assume that `ls-files` is called correctly but in the line
>   ```
>     run(['git', 'ls-files', '-z', '--stage', '--cached'], protocol_class=GitLsFilesProtocol...
>   ```
>     
>   I cannot immediately see this requirement/implication.
> * it seems we should/could aim for assembling a library of such common protocols

The, now-closed, big-runner PR with data_processors has a protocol, i.e. `StdOutErrCaptureProcessingGeneratorProtocol`, that passes all data through a data processor pipeline. The pipeline is configured by the protocol keyword argument `processors`.

With this concept, the `iter_gitworktree` call would look something like this:

```python
with run(
    ['git', 'ls-files', '-z', '--stage', '--cached'],
    protocol_class=StdOutCaptureProcessingGeneratorProtocol
    protocol_kwargs={'processors': [decode_processor('utf-8'), splitlines_processor('\x00', False), gititem_processor]}
) as git_ls_files:
    pass
```

That could of course be encapsulated into a specialized function, reducing the call to something like this

```python
with git_ls_files(path=path, ...) as gls:
    ...
```